File: api/src/main/java/dev/obscuria/fragmentum/client/ClientRegistrar.java
package dev.obscuria.fragmentum.client;

import dev.obscuria.fragmentum.content.registry.*;
import net.minecraft.client.model.geom.ModelLayerLocation;
import net.minecraft.client.model.geom.builders.LayerDefinition;
import net.minecraft.client.particle.ParticleProvider;
import net.minecraft.client.particle.SpriteSet;
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider;
import net.minecraft.client.renderer.blockentity.state.BlockEntityRenderState;
import net.minecraft.client.renderer.entity.EntityRendererProvider;
import net.minecraft.core.BlockPos;
import net.minecraft.core.particles.ParticleOptions;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.level.BlockAndTintGetter;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import org.checkerframework.checker.nullness.qual.Nullable;

@SuppressWarnings("unused")
public interface ClientRegistrar {

    <T extends Entity> void registerEntityRenderer(DeferredEntity<T> entity, EntityRendererProvider<T> provider);

    <T extends BlockEntity, S extends BlockEntityRenderState> void registerBlockEntityRenderer(DeferredBlockEntity<T> entity, BlockEntityRendererProvider<T, S> provider);

    <T extends ParticleOptions> void registerParticleRenderer(DeferredParticle<T> particle, ParticleProvider<T> provider);

    <T extends ParticleOptions> void registerTexturedParticleRenderer(DeferredParticle<T> particle, TexturedParticleProvider<T> provider);

    void registerModelLayer(ModelLayerLocation location, ModelLayerProvider provider);

    @SuppressWarnings("unchecked")
    void registerBlockColor(BlockColorProvider provider, Deferred<? extends Block>... blocks);

    @FunctionalInterface
    interface TexturedParticleProvider<T extends ParticleOptions> {

        ParticleProvider<T> create(SpriteSet spriteSet);
    }

    @FunctionalInterface
    interface ModelLayerProvider {

        LayerDefinition create();
    }

    @FunctionalInterface
    interface BlockColorProvider {

        int pick(BlockState state, @Nullable BlockAndTintGetter getter, @Nullable BlockPos pos, int layer);
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/api/client/ClientRegistrar.kt
package dev.obscuria.fragmentum.api.client

import dev.obscuria.fragmentum.api.registry.DeferredBlockEntity
import dev.obscuria.fragmentum.api.registry.DeferredEntity
import dev.obscuria.fragmentum.api.registry.DeferredItem
import dev.obscuria.fragmentum.api.registry.DeferredParticle
import net.minecraft.client.KeyMapping
import net.minecraft.client.model.geom.ModelLayerLocation
import net.minecraft.client.model.geom.builders.LayerDefinition
import net.minecraft.client.particle.ParticleProvider
import net.minecraft.client.particle.SpriteSet
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider
import net.minecraft.client.renderer.entity.EntityRendererProvider
import net.minecraft.client.renderer.item.ClampedItemPropertyFunction
import net.minecraft.core.particles.ParticleOptions
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.Entity
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.block.entity.BlockEntity

interface ClientRegistrar
{
    fun <T : Entity> registerEntityRenderer(
        type: DeferredEntity<T>,
        provider: EntityRendererProvider<T>
    )

    fun <T : BlockEntity> registerBlockEntityRenderer(
        type: DeferredBlockEntity<T>,
        provider: BlockEntityRendererProvider<T>
    )

    fun <T : ParticleOptions> registerParticleRenderer(
        type: DeferredParticle<T>,
        provider: ParticleProvider<T>
    )

    fun <T : ParticleOptions> registerTexturedParticleRenderer(
        type: DeferredParticle<T>,
        provider: TexturedParticleProvider<T>
    )

    fun registerModelLayer(
        layerLocation: ModelLayerLocation,
        provider: ModelLayerProvider
    )

    fun registerItemColor(
        provider: ItemColorProvider,
        vararg items: DeferredItem<*>
    )

    fun registerItemProperty(
        key: ResourceLocation,
        function: ClampedItemPropertyFunction
    )

    fun registerKeyMapping(keyMapping: KeyMapping)

    fun interface ModelLayerProvider
    {
        fun create(): LayerDefinition
    }

    fun interface TexturedParticleProvider<T : ParticleOptions>
    {
        fun create(spriteSet: SpriteSet): ParticleProvider<T>
    }

    fun interface ItemColorProvider
    {
        fun getColor(stack: ItemStack, layer: Int): Int
    }
}